src/Build/BackEnd/Shared/TargetResult.cs

```diff
@@ -284,6 +279,46 @@ private void RetrieveItemsFromCache()
             }
         }
 
+        private void TranslateItems(ITranslator translator)
+        {
+            var itemsCount = _items?.Length ?? 0;
+            translator.Translate(ref itemsCount);
+
+            if (translator.Mode == TranslationDirection.ReadFromStream)
+            {
+                LookasideStringInterner interner = new LookasideStringInterner(translator);
+                byte[] buffer = null;
+                translator.Translate(ref buffer);
+                ErrorUtilities.VerifyThrow(buffer != null, "Unexpected null items buffer during decompression.");
+
+                using MemoryStream itemsStream = new MemoryStream(buffer, 0, buffer.Length, writable: false, publiclyVisible: true);
+                var itemTranslator = BinaryTranslator.GetReadTranslator(itemsStream, null);
+
+                _items = new TaskItem[itemsCount];
+                for (int i = 0; i < _items.Length; i++)
+                {
+                    _items[i] = TaskItem.FactoryForDeserialization(itemTranslator, interner);
+                }
+            }
+            else
+            {
+                var interner = new LookasideStringInterner(StringComparer.Ordinal, _items.Length);
+                var capacity = itemsCount * 128; // Rough estimate 
+
+                using var itemsStream = new MemoryStream(capacity);
+                var itemTranslator = BinaryTranslator.GetWriteTranslator(itemsStream);
+
+                foreach (var taskItem in _items)
+                {
+                    taskItem.TranslateWithInterning(itemTranslator, interner);
+                }
+
+                interner.Translate(translator);
+                var itemsBuffer = itemsStream.ToArray();
```

Comment:
nit: It should be possible to eliminate this `byte[]` allocation. For example by calling `GetBuffer()` instead of `ToArray()` and passing the length of the array to `Translate` explicitly.

Reply:
Had to add new method on interface, to support this. Done